### PR TITLE
feat(e2e): add Maestro and basic test scenarios

### PR DESCRIPTION
**Pull request by araujobarret**
### Description

This PR adds Maestro and signup and login flows. The tests are working on Android and iOS.
There are no docs provided yet, I can add some but once we setup CI it's gonna change again, so I decided to skip it in this PR.

#### What is Maestro?

Maestro is an end-to-end tool([they](https://maestro.mobile.dev/) call themselves UI testing tool).
We use it in a similar way as other tools such as Cypress, Appium, and Detox in order to run tests over some defined scenarios.

In the follow-up PRs we'll have:
- CI
- documentation of how to run locally, CI
- a few more test scenarios

For those willing to test for now, remember to setup the **runtime**  environment variables
- MAESTRO_APP_ID: `net.artsy.artsy`(iOS) or `net.artsy.app`(Android)
- MAESTRO_TEST_EMAIL: found on 1Pass under `Maestro environment variables`
- MAESTRO_TEST_PASSWORD: found on 1Pass under `Maestro environment variables`

https://github.com/artsy/eigen/assets/15792853/130dfbda-c3d1-4361-bbb6-4b53d255c600

### PR Checklist

- [x] I have tested my changes on **iOS** and **Android**.
- [x] I hid my changes behind a **[feature flag]**, or they don't need one.
- [x] I have included **screenshots** or **videos**, or I have not changed the UI.
- [x] I have added **tests**, or my changes don't require any.
- [x] I added an **[app state migration]**, or my changes do not require one.
- [x] I have documented any **follow-up work** that this PR will require, or it does not require any.
- [x] I have added a **changelog entry** below, or my changes do not require one.

### To the reviewers 👀

- [ ] I would like **at least one** of the reviewers to **run** this PR on the simulator or device.

<details><summary>Changelog updates</summary>

#### Cross-platform user-facing changes

-

#### iOS user-facing changes

-

#### Android user-facing changes

-

#### Dev changes

- feat: add e2e maestro - araujobarret

</details>

Need help with something? Have a look at our [docs], or get in touch with us.

[app state migration]: ../blob/main/docs/adding_state_migrations.md
[feature flag]: ../blob/main/docs/developing_a_feature.md
[docs]: ../blob/main/docs/README.md

### CODE REVIEW

**Comment by ArtsyOpenSource** [comment]
### This PR contains the following changes:

- Dev changes (feat: add e2e maestro - araujobarret)
<p align="right">
  Generated by :no_entry_sign: <a href="https://danger.systems/js">dangerJS</a> against 253d4b6bcdd27958393dab883652b0401d9c6ec1
</p>

File: e2e/config.yml [hunk 1]
```diff
@@ -0,0 +1,9 @@
+appId: ${MAESTRO_APP_ID}
+---
+- runFlow:
+    file: flows/onboarding/login.yml
+- runFlow:
+    file: flows/onboarding/signupWithoutOnboarding.yml
+# This flow is broken because of performance issues and hangs of the Art Taste Quiz Screen
+# - runFlow:
+# file: flows/onboarding/onboardingQuiz.yml
```
Comment (gkartalis) [inline e2e/config.yml:9]: I will measure the performance there to see if we can improve stuff

File: e2e/flows/onboarding/onboardingQuiz.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+output.onboardingQuiz = { ...output.onboardingQuiz }
+
+if (!output.onboardingQuiz.button) {
+  output.onboardingQuiz = {
+    counter: 0,
+    button: () => {
+      console.log("counter", counter)
```
Comment (gkartalis) [inline e2e/flows/onboarding/onboardingQuiz.js:7]: Do we need to keep the log here? Should we hide it behind an `__DEV__` check if it is only for debugging?

File: src/app/Components/FancySwiper/FancySwiperIcons.tsx [hunk 3]
```diff
@@ -52,13 +52,13 @@ export const FancySwiperIcons = ({ swiper, OnPress }: FancySwiperIconsProps) =>
   return (
     <Flex flexDirection="row" justifyContent="space-around" pb={4} mx={4}>
       <Animated.View style={[{ flex: 1, alignItems: "center" }, dislikeIconAnimatedStyle]}>
-        <Touchable onPress={() => OnPress("left")}>
+        <Touchable onPress={() => OnPress("left")} accessibilityLabel="Skip">
```
Comment (gkartalis) [inline src/app/Components/FancySwiper/FancySwiperIcons.tsx:55]: super minor, but could we also add here `accessibilityRole="button"`?

File: src/app/Components/FancySwiper/FancySwiperIcons.tsx [hunk 4]
```diff
@@ -52,13 +52,13 @@ export const FancySwiperIcons = ({ swiper, OnPress }: FancySwiperIconsProps) =>
...
           <CloseIcon height={40} width={50} />
         </Touchable>
       </Animated.View>
       <Flex flex={1} alignItems="center">
         <Animated.View style={[likeIconAnimatedStyle, { position: "absolute" }]}>
-          <Touchable onPress={() => OnPress("right")}>
+          <Touchable onPress={() => OnPress("right")} accessibilityLabel="Like">
```
Comment (gkartalis) [inline src/app/Components/FancySwiper/FancySwiperIcons.tsx:61]: same as above - super minor, but could we also add here `accessibilityRole="button"`?

**Review by gkartalis — DISMISSED** [review]

File: e2e/flows/onboarding/onboardingQuiz.js (hunk 2 above)
Comment (araujobarret) [reply e2e/flows/onboarding/onboardingQuiz.js:7]: I will remove, since the test is broken for now this section needs to be re-visited

**Review by gkartalis — APPROVED** [review]
Looks great <3 Will try to run locally today to see if I have any issues with that

**Review by gkartalis — COMMENTED** [review]
Let's pair and document the steps we took to build run Maestro locally on my machine since I had a hard time running it

File: e2e/config.yml [hunk 5]
```diff
@@ -0,0 +1,9 @@
+appId: ${MAESTRO_APP_ID}
```
[the remark this answers is not in the document]
Comment (araujobarret) [reply e2e/config.yml:1]: we would need to setup some bash scripts, I'm fine doing it, just better than putting more stuff in the package.json scripts section

**Comment by damassi** [comment]
@araujobarret - can you add a link to the description for Maestro and a description of what it is?

**Comment by gkartalis** [comment]
## Steps to run


1. install maestro by following the instructions [here](https://maestro.mobile.dev/getting-started/installing-maestro)
     - `curl -Ls "https://get.maestro.mobile.dev" | bash`
     - `brew tap facebook/fb`
     - `brew install facebook/fb/idb-companion`

### iOS 

1. `CI=1 yarn pod-install && yarn maestro:ios:release:build` 
1. `yarn open-sim` open iOS simulator (after sim is booted run next step)
1. `yarn maestro:ios:release:install` 

1.  set the ENV vars `export MAESTRO_APP_ID=app_id` `export MAESTRO_TEST_EMAIL=email@email.com` `export MAESTRO_TEST_PASSWORD=password`
1. run the test that you want to run `maestro test e2e/config.yml`

### Android

1. `yarn maestro:android:release:build`
1. `yarn open-emulator` 
1. `yarn maestro:android:release:install` 

1.  set the ENV vars `export MAESTRO_APP_ID=app_id` `export MAESTRO_TEST_EMAIL=email@email.com` `export MAESTRO_TEST_PASSWORD=password`
1. run the test that you want to run `maestro test e2e/flows/onboarding/signup.yml/config.yml`

**Review by brainbicycle — APPROVED** [review]
looks great!

File: src/app/Scenes/Onboarding/OnboardingCreateAccount/TermsOfServiceCheckbox.tsx [hunk 6]
```diff
@@ -18,7 +18,21 @@ export const TermsOfServiceCheckbox: React.FC<TermsOfServiceCheckboxProps> = ({
...
+          checkboxAccessibilityProps={{
+            accessible: true,
+            accessibilityRole: "checkbox",
+            accessibilityLabel: "checkbox of consent",
+            accessibilityHint: "Check this element to consent with the Terms of Service",
+            accessibilityState: {
+              checked,
+            },
+          }}
+        >
```
Comment (MounirDhahri) [inline src/app/Scenes/Onboarding/OnboardingCreateAccount/TermsOfServiceCheckbox.tsx:35]: 👍

File: package.json [hunk 7]
```diff
@@ -49,6 +49,11 @@
     "local-palette-dev:stop": "./scripts/palette/yalc-unlink-local-palette",
     "lint": "eslint --cache --cache-location '.cache/eslint/' --ext .ts,.tsx --fix",
     "lint:all": "yarn lint .",
+    "maestro:android:release:build": "cd android; ./gradlew bundleRelease; cd ..",
+    "maestro:android:release:install": "cd android; ./gradlew installRelease; cd ..",
+    "maestro:ios:release:build": "xcodebuild -workspace ios/Artsy.xcworkspace -scheme Artsy -configuration Store -sdk iphonesimulator -derivedDataPath ios/build",
+    "maestro:ios:release:install": "xcrun simctl install booted ios/build/Build/Products/Store-iphonesimulator/Artsy.app",
+    "open-emulator": "emulator -avd $(emulator -list-avds | head -1) &",
```
Comment (MounirDhahri) [inline package.json:56]: this is a nice command - is emulator an alias for something?

**Review by MounirDhahri — COMMENTED** [review]
LGTM overall but just a few comments:
- Is it possible to move flows to be along with the files as we do for our other tests?
- I believe we need to update the setup/install scripts to install maestro or have a separate script for setting up e2e
- I am not sure if `maestro:android:release:build`, `maestro:android:release:install`, `maestro:ios:release:build` and `maestro:android:release:install` belong all to `package.json`. I would be in favor of having a script for android and one for iOS that builds then installs the app. Such a script can be called from`setup:e2e:android` and `setup:e2e:iOS`. Maybe it can even check if the maestro command is available and if not prompt the dev to install it.
- Related to the above, `maestro test [test path]` seems to me as a good command to have in the `package.json`

File: package.json (hunk 7 above)
Comment (araujobarret) [reply package.json:56]: it's from adb-tools `emulator`, you probably have it or just haven't setup all env vars necessary

**Comment by araujobarret** [comment]
Answering you @MounirDhahri

> Is it possible to move flows to be along with the files as we do for our other tests?

Don't think it belongs there, Maestro is living together in Eigen to be like a parasite to use the benefits of building and using the generated builds, it's an independent application.

> I believe we need to update the setup/install scripts to install maestro or have a separate script for setting up e2e

The idea is to provide Docs on how to install but we should point to their official docs in case something changes there people would just use up-to-date information, but I will add the notes we sorted in @gkartalis' comment
I will also add some sh file to initiate the test suite too

> I am not sure if maestro:android:release:build, maestro:android:release:install, maestro:ios:release:build and maestro:android:release:install belong all to package.json. I would be in favor of having a script for android and one for iOS that builds then installs the app. Such a script can be called fromsetup:e2e:android and setup:e2e:iOS. Maybe it can even check if the maestro command is available and if not prompt the dev to install it.

I had the same feeling, specially because it's an independent application, I will move it to some /scripts section. Do you agree @gkartalis ?

> Related to the above, maestro test [test path] seems to me as a good command to have in the package.json

Maestro is not installed in the package.json module if we add the command there yarn is going to try to find something under node_modules. Moreover, if we split the install, and build steps into the shell would make sense for the test command to be there too, what do you think?

File: package.json (hunk 7 above)
Comment (MounirDhahri) [reply package.json:56]: aha - i do have it, `yarn open-emulator` though is not working for me. I thought it's an npm package or something like that

Comment (araujobarret) [reply package.json:56]: interesting, what do you see if you type in your terminal `emulator -list-avds | head -1`?

Comment (MounirDhahri) [reply package.json:56]: adb emulator work for me - it's running the command with yarn that breaks for me

Comment (araujobarret) [reply package.json:56]: ah my bad, you need to be able to run it without yarn, just `emulator -list-avds` for example, if that isn't working then you're missing `platform-tools` in your PATH variable.
Can you check if you have something like:
```
export ANDROID_HOME=$HOME/Library/Android/sdk
export ANDROID_SDK_ROOT=$HOME/Library/Android/sdk
export PATH=$PATH:$ANDROID_SDK_ROOT/emulator
export PATH=$PATH:$ANDROID_SDK_ROOT/tools
export PATH=$PATH:$ANDROID_SDK_ROOT/tools/bin
export PATH=$PATH:$ANDROID_SDK_ROOT/platform-tools
```
in your .zshrc

**Comment by gkartalis** [comment]
> > I am not sure if maestro:android:release:build, maestro:android:release:install, maestro:ios:release:build and maestro:android:release:install belong all to package.json. I would be in favor of having a script for android and one for iOS that builds then installs the app. Such a script can be called fromsetup:e2e:android and setup:e2e:iOS. Maybe it can even check if the maestro command is available and if not prompt the dev to install it.

> I had the same feeling, specially because it's an independent application, I will move it to some /scripts section. Do you agree @gkartalis ?

I think I agree with this yes we can have them as scripts

> > Related to the above, maestro test [test path] seems to me as a good command to have in the package.json

> Maestro is not installed in the package.json module if we add the command there yarn is going to try to find something under node_modules. Moreover, if we split the install, and build steps into the shell would make sense for the test command to be there too, what do you think?

I think this should live on the scripts and we can add a section on our gettingstarted.md docs to install this `maestro` dep

**Comment by araujobarret** [comment]
thanks @gkartalis and @MounirDhahri I will address the requests tomorrow so we can move on 👍

File: package.json (hunk 7 above)
Comment (MounirDhahri) [reply package.json:56]: I have emulator and platform tools, I miss some that I am not usually using

**Comment by MounirDhahri** [comment]
Thanks @araujobarret - my concerns can be addressed in a separate PR if needed no need to have them as part of the ticket as long as we don't forget to come back to them

**Comment by araujobarret** [comment]
Alright, I added some basic instructions we elaborated with @gkartalis and will move the setup scripts to `/scripts` in the upcoming CI PR.

**Pull request merged**